c++/src/kj/source-location.h

```diff
@@ -23,20 +23,33 @@
 
 #include "string.h"
 
-#if __clang__ && __clang_major__ >= 9
-#define KJ_CALLER_COLUMN() __builtin_COLUMN()
 // GCC does not implement __builtin_COLUMN() as that's non-standard but MSVC & clang do.
 // MSVC does as of version https://github.com/microsoft/STL/issues/54) but there's currently not any
 // pressing need for this for MSVC & writing the write compiler version check is annoying.
+// Checking for clang version is problematic due to the way that XCode lies about __clang_major__.
+// Instead we use __has_builtin as the feature check to check clang.
+// Context: https://github.com/capnproto/capnproto/issues/1305
+#ifdef __has_builtin
+#if __has_builtin(__builtin_COLUMN)
+#define KJ_CALLER_COLUMN() __builtin_COLUMN()
+#else
+#define KJ_CALLER_COLUMN() 0
+#endif
 #else
 #define KJ_CALLER_COLUMN() 0
 #endif
 
 #if __cplusplus > 201703L
 #define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
-#elif __clang__ && __clang_major__ >= 9
+#elif defined(__has_builtin)
 // Clang 9 added these builtins: https://releases.llvm.org/9.0.0/tools/clang/docs/LanguageExtensions.html
-#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
+// Use __has_builtin as the way to detect this because __clang_major__ is unreliable (see above
+// about issue with Xcode-provided clang).
+#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION (   \
```

Comment:
Missed the `(` `)` protection around the macro. Otherwise `!KJ_COMPILER_SUPPORTS_SOURCE_LOCATION` expands to the wrong boolean.